internal/registry/registry.go

```diff
@@ -0,0 +1,52 @@
+package registry
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+)
+
+// A declaratively registered function
+type RegisteredFunction struct {
```

Comment:
Letting users choose a name for the function to register is a bit risky. In case of a name conflict, we'll have to decide whether to fail the registration or which function to run. Having `CloudEventFn` and `HTTPFn` with one guaranteed to be empty is a bit odd to me.

I'm leaning towards keeping this as simple as possible to avoid possible name collisions and type conflicts. Is it possible that we only store the function and its type here?
Then user can do `funcframework.HTTP(HelloWorld)`

What do you think?

Reply:
This is a good comment, and is one thing that I think depends on per language. I'll reply to the points made here individually.

- Choosing a name for the function
  -  In some languages, I believe it is required (or idiomatic) to choose a name for a function and not use reflection. For example, this is now Node registers functions (https://github.com/GoogleCloudPlatform/functions-framework-nodejs/pull/347). In python, decorators make sense. In Go, I believe this register function makes sense.
- Having `CloudEventFn` or `HTTPFn` be empty
  -  In Go, given there are no union types, it is common for struct to have fields that are empty. I believe this is the best way to structure the function registry.

Allowing a user to register a function without a name seems fine, like `funcframework.HTTP(HelloWorld)`. I would think that this would be an additional function, as some use-cases may still want a named function (for calling other functions, multiple functions).

WDYT @anniefu @matthewrobertson ?

Reply:
The implementation should match what Annie talked about. Will resolve! Feel free to open / comment.